.kokoro/trampoline_v2.sh

```diff
@@ -0,0 +1,352 @@
+
+log_yellow "Start pulling the Docker image: ${TRAMPOLINE_IMAGE}."
+if docker pull "${TRAMPOLINE_IMAGE}"; then
+    log_green "Finished pulling the Docker image: ${TRAMPOLINE_IMAGE}."
+    has_cache="true"
+else
+    log_red "Failed pulling the Docker image: ${TRAMPOLINE_IMAGE}."
+    has_cache="false"
+fi
+
+
+# The default user for a Docker container has uid 0 (root). To avoid
+# creating root-owned files in the build directory we tell docker to
+# use the current user ID.
+user_uid="$(id -u)"
+user_gid="$(id -g)"
+user_name="$(id -un)"
+
+# To allow docker in docker, we add the user to the docker group in
+# the host os.
+docker_gid=$(cut -d: -f3 < <(getent group docker))
+
+update_cache="false"
+if [[ "${TRAMPOLINE_DOCKERFILE:-none}" != "none" ]]; then
+    # Build the Docker image from the source.
+    context_dir=$(dirname "${TRAMPOLINE_DOCKERFILE}")
+    docker_build_flags=(
+	"-f" "${TRAMPOLINE_DOCKERFILE}"
+	"-t" "${TRAMPOLINE_IMAGE}"
+	"--build-arg" "UID=${user_uid}"
+	"--build-arg" "GID=${user_gid}"
+	"--build-arg" "USERNAME=${user_name}"
+	"--build-arg" "DOCKER_GID=${docker_gid}"
+    )
+    if [[ "${has_cache}" == "true" ]]; then
+	docker_build_flags+=("--cache-from" "${TRAMPOLINE_IMAGE}")
+    fi
+
+    log_yellow "Start building the docker image."
+    if [[ "${TRAMPOLINE_SHOW_COMMAND:-false}" == "true" ]]; then
+	echo "docker build" "${docker_build_flags[@]}" "${context_dir}"
+    fi
+    if docker build "${docker_build_flags[@]}" "${context_dir}"; then
+	log_green "Finished building the docker image."
+	update_cache="true"
+    else
+	log_red "Failed to build the Docker image. Aborting."
+	exit 1
+    fi
+else
+    if [[ "${has_cache}" != "true" ]]; then
+	log_red "failed to download the image ${TRAMPOLINE_IMAGE}, aborting."
+	exit 1
+    fi
+fi
+
+# We use an array for the flags so they are easier to document.
+docker_flags=(
+    # Remove the container after it exists.
+    "--rm"
+
+    # Use the host network.
+    "--network=host"
+
+    # Run in priviledged mode. We are not using docker for sandboxing or
+    # isolation, just for packaging our dev tools.
+    "--privileged"
+
+    # Run the docker script with the user id. Because the docker image gets to
+    # write in ${PWD} you typically want this to be your user id.
+    # Also to allow docker in docker, we use docker gid on the host.
+    "--user" "${user_uid}:${docker_gid}"
+
+    # Pass down the USER.
+    "--env" "USER=${user_name}"
+
+    # Mount the project directory inside the Docker container.
+    "--volume" "${PROJECT_ROOT}:${TRAMPOLINE_WORKSPACE}"
+    "--workdir" "${TRAMPOLINE_WORKSPACE}"
+    "--env" "PROJECT_ROOT=${TRAMPOLINE_WORKSPACE}"
+
+    # Mount the temporary home directory.
+    "--volume" "${tmphome}:/h"
+    "--env" "HOME=/h"
+
+    # Allow docker in docker.
+    "--volume" "/var/run/docker.sock:/var/run/docker.sock"
+
+    # Mount the /tmp so that docker in docker can mount the files
+    # there correctly.
+    "--volume" "/tmp:/tmp"
```

Comment:
Are you sure docker in docker needs `/tmp`? I thought it only needed the socket?

Reply:
If you're using docker in docker and try to mount files or directories inside the container, the path must be the same location on the docker host too.

It worked with trampoline V1 because the path was the same (`/tmpfs/src/github/python-docs-samples`).

But now we want to mount the project at `/workspace`, we need a path which is guaranteed to be the same on the docker host.

Reply:
Ah ok, I think I understand. 

I think you can use a named volume and mount it to both containers to avoid exposing the hosts `/tmp` folder. I'm not sure if this is beneficial or not - my instinct is that exposing the entire `/tmp` folder might not be a great idea, but I'm not sure.